### ISSUE TITLE
Async curry?

### ISSUE THREAD

**Issue by dbfreem** [ethereum/web3.py#2583]
I was starting to look at ways to async `name_to_address` middleware now that `AsyncENS` is done. I ran into a problem that I wanted to get some feedback on. I ran into this [curry](https://github.com/ethereum/web3.py/blob/master/web3/_utils/normalizers.py#L207). Then I realized that I really need curry in this example because I want to pass in the w3 instance when the middleware is added so it can be used to resolve the name of the address. When I try to use curry on an async method I get and exception.

Has anyone come up with a solution for async curry methods already while writing some of the other async code? If not I will keep thinking about a good way to make this work.

**Comment by kclowes** [comment]
We've talked in the past about using the [paco library](https://github.com/h2non/paco), but it looks pretty stale and may or may not be maintained. I wonder if we could take the `curry` code from [here](https://github.com/h2non/paco/blob/master/paco/curry.py) and apply it? I haven't looked at the code very closely tbh so I'm not sure this will be possible/what we want.

**Comment by pacrob** [comment]
I was looking at the async curry yesterday. I'm not sure if it will fit in the rest of the flow, but you can use it by only `await`ing the final step, like:

``` py
>>> async_curry = async_abi_ens_resolver(w3)
>>> async_curry = async_curry('tuple')
>>> async_curry = async_curry((2, 2))
>>> await curry
('tuple', (2, 2))
```

**Comment by dbfreem** [comment]
So I took a look at paco and I agree it looks unmaintained. I also looked at several other async functional libraries and none of them really seemed to be maintained which is unfortunate. Not sure which way you all want to take it from here on the functional front. I agree with the functional approach in the library so I don't think it would be a good idea to rewrite it in another paradigm just because there isn't an async library. 

I stopped working on the names middleware because it was heavily dependent on the underlying curry functions.

Thoughts??
